build: prepare for TypeScript 3.9

alan-agius4 opened the pull request:
- Fix several compilation errors
- Update @microsoft/api-extractor to be compatible with TypeScript 3.9

alan-agius4 commented on packages/service-worker/BUILD.bazel:
```diff
@@ -59,12 +59,14 @@ ng_package(
-    ],
-    goldenDir = "angular/goldens/public-api/service-worker",
-)
+# Currently this fails with 'Error: Symbol "Generator_2" was aliased as "Generator". Aliases are not supported."' from ts-api-guardian
```
I am also seeing the same problem for other API tests.

petebacondarwin committed:
fix(compiler-cli): handle FileSystem case-sensitivity is correctly

The `NodeJSFileSystem.isCaseSensitive()` method was computing
completely the opposite value for its result.

Also the `NgtscCompilerHost.getCanonicalFileName()` method was
not actually calling `useCaseSensitiveFileNames()` so it was always
assuming that the filenames were case sensitive and not actually
canonicalizing them. Instead just returning the original.

Once these two bugs were fixed, a number of issues with the rest of the
handling of case-insensitive files in compiler-cli and tests were uncovered.

This commit also fixes these issues. Here are some further notes:

The TS compiler canonicalizes filenames internally for lookups etc. So
if we are also doing file lookups we need to use the same canonicalized
forms. This is achieved by injecting the `ts.CompilerHost` where needed
and calling `getCanonicalFileName()`. For example, when computing the
`rootDirs` in the `LogicalFileSystem`.

Some tests were checking that particular source files were being exposed
by comparing the `ts.SourceFile.fileName` property with some non-canonical
filename. In some cases this would fail. Instead these comparisons are
achieved by directly comparing the SourceFile objects, if known, or getting
the SourceFile object from a non-canonical filename from the program, by
calling `program.getSourceFile(nonCanonicalFileName)`.

The `MockFileSystemWindows` class was not considering lower case drive names
when computing if a path is rooted (i.e. absolute).

The `NodeJSFileSystem.isCaseSensitive()` method was relying upon its `exists()`
method. But this itself relies upon the `isCaseSensitive()` to determine
if a file exists. So there was an infinite recursion. This is solved by
actually using the underlying real `fs.existsSync` to compute whether the
file system is case sensitive.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/file_system/testing/src/test_helper.ts:
```diff
@@ -75,7 +75,7 @@ export function initMockFileSystem(os: string, cwd?: AbsoluteFsPath): void {
   switch (os) {
     case 'OS/X':
-      return new MockFileSystemPosix(/* isCaseSensitive */ false, cwd);
+      return new MockFileSystemPosix(/* isCaseSensitive */ true, cwd);
```
This is not correct. OS/X by default is case-insensitive, although it doesn't have to be.
In any case, changing this to true, makes it identical to the `Unix` FS below, which doesn't add any value.

alan-agius4 committed:
build: prepare for TypeScript 3.9

- Fix several compilation errors
- Update @microsoft/api-extractor to be compatible with TypeScript 3.9

alan-agius4 committed:
build: pin @microsoft/api-extractor to 7.7.11

Version 7.7.12 generates aliases for objects named the same as global symbols which breaks ts-api-guardian since it doesn't support aliases.

See: https://github.com/microsoft/rushstack/issues/1830

alan-agius4 commented on goldens/public-api/core/core.d.ts:
```diff
@@ -313,7 +313,7 @@ export declare class ErrorHandler {
 }
 
-export declare class EventEmitter<T extends any> extends Subject<T> {
+export declare class EventEmitter<T = any> extends Subject<T> {
```
This is done, so that we don't have a breaking change. Because in TS 3.9 `T extends any` behaviour changed

```
function foo<T extends any>(arg: T) {
    arg.spfjgerijghoied; // no error!
}

function foo<T extends any>(arg: T) {
    arg.spfjgerijghoied;
    //  ~~~~~~~~~~~~~~~
    // Property 'spfjgerijghoied' does not exist on type 'T'.
}
```

`T extends any` is more or less the same as `T extends unknown`

gkalpak commented on packages/compiler-cli/ngcc/index.ts:
```diff
@@ -8,15 +8,15 @@
-export function process(options: AsyncNgccOptions): Promise<void>;
-export function process(options: SyncNgccOptions): void;
-export function process(options: NgccOptions): void|Promise<void> {
+export function process<T extends AsyncNgccOptions|SyncNgccOptions>(options: T):
+    T extends AsyncNgccOptions ? Promise<void>: void;
+export function process(options: AsyncNgccOptions|SyncNgccOptions): void|Promise<void> {
```
IMO, this would be more readable as:

```suggestion
export function process<T extends NgccOptions>(options: T):
    T extends AsyncNgccOptions ? Promise<void>: void {
```

Same for `mainNgcc()`. Did I miss something?

alan-agius4 replied:
There are not the same.

https://www.typescriptlang.org/v2/en/play?ts=3.9.0-beta#code/GYVwdgxgLglg9mABAUwB7IiKyAUBKALkQDc4YATAHwAUAnOAWxgGdkAeUigPkQG8AofgF9BaAA5xaURDDDZawAIYRkiAMoBPSADkA5hAgB5MbATM+-RFcSLmWiAH4iSgDatriYaNQSpMucgKyqoAgnY6+kYm8GDmAh629kRQtCCqVl78APRZiADqkgDWzPziktKgkKZIYvQqzMxsACooqNhg5OZh9noGxtXMlJoRfdFmXDhwY7FETYSILWjtnYjdI1EDiA6IdIws7JzkXESHANylPuWIldAxiLVw9cyT08xEaxC9GzGDw5+R-R+80ONHoTFYHDIRwsHloyCgIFoSDQGCwuDw5xE2VyAFFaPRaBdfBVwLcEPc6sgGgB9ZqtZZdcL-UYDIY9AGvCZTAazeaLNrIDqM9ksn5bHZg-aQ7gnKEw6xwhFI1qo7D4TGCIA

gkalpak replied:
Hm...it would work for `process()` is `mainNgcc()` was also updated to the same signature (as I suggested above). But I now realize that we would need to make changes in `mainNgcc()` internally to satisfy the return new type.

So, let's keep it as is :grin:

alan-agius4 commented:
There seems to be 2 main pending failure points.

**NGCC**
- @petebacondarwin, offered his help to find the root cause and solution

**LS**
- Found the root cause and will laise with @kyliau

petebacondarwin committed:
fixup! fix(compiler-cli): handle FileSystem case-sensitivity is correctly

Commit:
fix(language-service): use empty statement as parent of type node

Commit:
test(language-service): external template update should not invalidate program

alan-agius4 committed:
fixup! build: prepare for TypeScript 3.9

alan-agius4 commented:
Closing in favour of https://github.com/angular/angular/pull/36989
